core/lib/zksync_core/src/genesis.rs

```diff
@@ -120,29 +202,56 @@ pub async fn ensure_genesis_state(
         rollup_last_leaf_index,
     )
     .await?;
-    tracing::info!("operations_schema_genesis is complete");
-
     transaction.commit().await?;
+    Ok((
+        genesis_root_hash,
+        block_commitment.hash().commitment,
+        rollup_last_leaf_index,
+    ))
+}
 
-    // We need to `println` this value because it will be used to initialize the smart contract.
-    println!("CONTRACTS_GENESIS_ROOT={:?}", genesis_root_hash);
-    println!(
-        "CONTRACTS_GENESIS_BATCH_COMMITMENT={:?}",
-        block_commitment.hash().commitment
-    );
-    println!(
-        "CONTRACTS_GENESIS_ROLLUP_LEAF_INDEX={}",
-        rollup_last_leaf_index
-    );
-    println!(
-        "CHAIN_STATE_KEEPER_BOOTLOADER_HASH={:?}",
-        base_system_contracts_hashes.bootloader
-    );
-    println!(
-        "CHAIN_STATE_KEEPER_DEFAULT_AA_HASH={:?}",
-        base_system_contracts_hashes.default_aa
-    );
+pub async fn ensure_genesis_state(
+    storage: &mut StorageProcessor<'_>,
+    genesis_params: &GenesisParams,
+) -> Result<H256, GenesisError> {
+    let mut transaction = storage.start_transaction().await?;
+
+    if !transaction.blocks_dal().is_genesis_needed().await? {
+        tracing::debug!("genesis is not needed!");
+        return Ok(transaction
+            .blocks_dal()
+            .get_l1_batch_state_root(L1BatchNumber(0))
+            .await
+            .context("failed fetching state root hash for genesis L1 batch")?
+            .context("genesis L1 batch hash is empty")?);
+    }
+
+    tracing::info!("running regenesis");
+    let (genesis_root_hash, commitment, rollup_last_leaf_index) =
+        insert_genesis_batch(&mut transaction, genesis_params).await?;
+    if genesis_params.config.genesis_root_hash != genesis_root_hash {
```

Comment:
Why do we even need `genesis_params.config.genesis_root_hash`. Is it used anywhere else except this place? If not, do we need this check?

Reply:
Yes, it's used during contract deployment in this way we can deploy contracts without building the server

Reply:
So, when genesis root hash/commitment changes (after another protocol upgrade) expected flow will be:
run zk init -> it fails -> change variable -> run zk init again? If so then it doesn't look good. I don't mind merging it because the current approach with printing is also bad but would be nice if we could improve it

Reply:
Yep, we can create a method for just updating Genesis config.